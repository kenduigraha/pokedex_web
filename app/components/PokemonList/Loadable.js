/**
 *
 * Asynchronously loads the component for PokemonList
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
