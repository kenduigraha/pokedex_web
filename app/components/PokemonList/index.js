/**
 *
 * PokemonList
 *
 */

import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Row } from 'antd';
import CharactersCard from 'components/CharactersCard';
// import styled from 'styled-components';

// import { FormattedMessage } from 'react-intl';
// import messages from './messages';

class PokemonList extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      limit: 20,
      offset: 0,
      cardLoading: false,
    };
  }

  /**
   * hit API pokemon list
   */
  fetchDataPokemonList = params => {
    this.props.getPokemonList(params);
  };

  componentWillReceiveProps(nextProps) {
    document.addEventListener('scroll', this.trackScrolling);

    if (nextProps.pokemonList.infinity === false) {
      this.setState({ cardLoading: false });
    }
  }

  /**
   * remove event listener scroll
   */
  componentWillUnmount() {
    document.removeEventListener('scroll', this.trackScrolling);
  }

  /**
   * check reached bottom or not
   * @param {element} el
   */
  isBottom(el) {
    if (el) return el.getBoundingClientRect().bottom <= window.innerHeight;
    return false;
  }

  /**
   * track scroll & hit API after reached
   */
  trackScrolling = () => {
    const wrappedElement = document.getElementById('content');

    if (
      this.props.pokemonList.isLoading === false &&
      this.isBottom(wrappedElement)
    ) {
      this.props.updateFlagInfinity(true);

      document.removeEventListener('scroll', this.trackScrolling);

      if (this.props.pokemonList.infinity === true) {
        this.setState(
          prevState => ({
            offset: prevState.offset + 20,
          }),
          () => {
            this.fetchDataPokemonList({
              limit: this.state.limit,
              offset: this.state.offset,
            });
          },
        );
      }
    }
  };

  render() {
    const { pokemonList } = this.props;
    return (
      <div id="content" style={{ background: '#ECECEC', padding: '30px' }}>
        {pokemonList.infinity === false && pokemonList.isLoading === true ? (
          'loading'
        ) : (
          <div>
            <Row gutter={24}>
              {pokemonList.data.map(char => (
                <CharactersCard
                  key={`${char.name}`}
                  loading={this.state.cardLoading}
                  data={char}
                />
              ))}
            </Row>
            {pokemonList.infinity === true && 'loading'}
          </div>
        )}
      </div>
    );
  }
}

PokemonList.propTypes = {
  pokemonList: PropTypes.object.isRequired,
  getPokemonList: PropTypes.func.isRequired,
  updateFlagInfinity: PropTypes.func.isRequired,
};

export default PokemonList;
