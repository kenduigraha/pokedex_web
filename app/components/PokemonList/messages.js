/*
 * PokemonList Messages
 *
 * This contains all the text for the PokemonList component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.PokemonList';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'This is the PokemonList component!',
  },
});
