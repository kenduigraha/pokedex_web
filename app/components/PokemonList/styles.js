import styled from 'styled-components';
import { Button } from 'antd';

const PokemonListDiv = styled.div`
  padding: 30px;
`;

const PokemonTypeButton = styled(Button)`
  text-transform: uppercase;
  margin-right: 5px;
  min-width: 100px;
  padding: 5px 10px;
  color: white !important;

  &.btn-type-normal {
    background: #a8a878;
  }

  &.btn-type-fire {
    background: #f08030;
  }

  &.btn-type-fighting {
    background: #c03028;
  }

  &.btn-type-water {
    background: #6890f0;
  }

  &.btn-type-flying {
    background: #a890f0;
  }

  &.btn-type-grass {
    background: #78c850;
  }

  &.btn-type-poison {
    background: #a040a0;
  }

  &.btn-type-electric {
    background: #f8d030;
  }

  &.btn-type-ground {
    background: #e0c068;
  }

  &.btn-type-psychic {
    background: #f85888;
  }

  &.btn-type-rock {
    background: #b8a038;
  }

  &.btn-type-ice {
    background: #98d8d8;
  }

  &.btn-type-bug {
    background: #a8b820;
  }

  &.btn-type-dragon {
    background: #7038f8;
  }

  &.btn-type-ghost {
    background: #705898;
  }

  &.btn-type-dark {
    background: #705848;
  }

  &.btn-type-steel {
    background: #b8b8d0;
  }

  &.btn-type-fairy {
    background: #ee99ac;
  }
`;

export { PokemonListDiv, PokemonTypeButton };
