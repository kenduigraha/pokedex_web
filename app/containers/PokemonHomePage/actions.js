/*
 *
 * PokemonHomePage actions
 *
 */

import {
  DEFAULT_ACTION,
  GET_POKEMON_LIST_START,
  GET_POKEMON_LIST_SUCCESS,
  GET_POKEMON_LIST_FAILED,
  UPDATE_FLAG_INFINITY,
  GET_POKEMON_DETAIL_START,
  GET_POKEMON_DETAIL_SUCCESS,
  GET_POKEMON_DETAIL_FAILED,
  GET_POKEMON_TYPES_START,
  GET_POKEMON_TYPES_SUCCESS,
  GET_POKEMON_TYPES_FAILED,
} from './constants';

export function defaultAction() {
  return {
    type: DEFAULT_ACTION,
  };
}

export function getPokemonListStart(params) {
  return {
    type: GET_POKEMON_LIST_START,
    params,
  };
}

export function getPokemonListSuccess(data) {
  return {
    type: GET_POKEMON_LIST_SUCCESS,
    data,
  };
}

export function getPokemonListFailed(error) {
  return {
    type: GET_POKEMON_LIST_FAILED,
    error,
  };
}

export function updateFlagInfinityStart(flag) {
  return {
    type: UPDATE_FLAG_INFINITY,
    data: flag,
  };
}

export function getPokemonDetailStart(data) {
  return {
    type: GET_POKEMON_DETAIL_START,
    data,
  };
}

export function getPokemonDetailSuccess(data) {
  return {
    type: GET_POKEMON_DETAIL_SUCCESS,
    data,
  };
}

export function getPokemonDetailFailed(error) {
  return {
    type: GET_POKEMON_DETAIL_FAILED,
    error,
  };
}

export function getPokemonTypesListStart(params) {
  return {
    type: GET_POKEMON_TYPES_START,
    params,
  };
}

export function getPokemonTypesListSuccess(data) {
  return {
    type: GET_POKEMON_TYPES_SUCCESS,
    data,
  };
}

export function getPokemonTypesListFailed(error) {
  return {
    type: GET_POKEMON_TYPES_FAILED,
    error,
  };
}
