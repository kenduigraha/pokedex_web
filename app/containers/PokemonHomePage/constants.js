/*
 *
 * PokemonHomePage constants
 *
 */

export const DEFAULT_ACTION = 'app/PokemonHomePage/DEFAULT_ACTION';

export const GET_POKEMON_LIST_START =
  'app/PokemonHomePage/GET_POKEMON_LIST_START';
export const GET_POKEMON_LIST_SUCCESS =
  'app/PokemonHomePage/GET_POKEMON_LIST_SUCCESS';
export const GET_POKEMON_LIST_FAILED =
  'app/PokemonHomePage/GET_POKEMON_LIST_FAILED';
export const UPDATE_FLAG_INFINITY = 'app/PokemonHomePage/UPDATE_FLAG_INFINITY';
