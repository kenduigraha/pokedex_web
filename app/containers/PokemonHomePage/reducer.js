/*
 *
 * PokemonHomePage reducer
 *
 */
import produce from 'immer';
import {
  DEFAULT_ACTION,
  GET_POKEMON_LIST_START,
  GET_POKEMON_LIST_SUCCESS,
  GET_POKEMON_LIST_FAILED,
  UPDATE_FLAG_INFINITY,
  GET_POKEMON_DETAIL_START,
  GET_POKEMON_DETAIL_SUCCESS,
  GET_POKEMON_DETAIL_FAILED,
  GET_POKEMON_TYPES_START,
  GET_POKEMON_TYPES_SUCCESS,
  GET_POKEMON_TYPES_FAILED,
  GET_POKEMON_LIST_BY_TYPES_START,
  GET_POKEMON_LIST_BY_TYPES_SUCCESS,
  GET_POKEMON_LIST_BY_TYPES_FAILED,
} from './constants';

export const initialState = {
  pokemonList: {
    data: [],
    error: {},
    isLoading: false,
    infinity: false,
  },
  pokemonDetail: {
    data: {},
    error: {},
    isLoading: false,
  },
  pokemonTypes: {
    data: [],
    error: {},
    isLoading: false,
  },
};

/* eslint-disable default-case, no-param-reassign */
const pokemonHomePageReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case DEFAULT_ACTION:
        break;
      case GET_POKEMON_LIST_BY_TYPES_START:
      case GET_POKEMON_LIST_START:
        draft.pokemonList.isLoading = true;
        break;
      case GET_POKEMON_LIST_SUCCESS:
        draft.pokemonList.isLoading = false;
        if (!action.data.results) {
          draft.pokemonList.data = [action.data.species];
        } else {
          const results =
            state.pokemonList.data.length === 1
              ? action.data.results
              : state.pokemonList.data.concat(action.data.results);

          draft.pokemonList.data = action.data.results ? results : [];
        }
        break;

      case GET_POKEMON_LIST_BY_TYPES_SUCCESS:
        draft.pokemonList.isLoading = false;
        draft.pokemonList.data = action.data.pokemon.map(detail => ({
          ...detail.pokemon,
        }));
        break;
      case GET_POKEMON_LIST_BY_TYPES_FAILED:
      case GET_POKEMON_LIST_FAILED:
        draft.pokemonList.isLoading = false;
        draft.pokemonList.infinity = false;
        draft.pokemonList.data = [];
        draft.pokemonList.error = action.error;
        break;
      case UPDATE_FLAG_INFINITY:
        draft.pokemonList.infinity = action.data;
        break;

      // GET DETAIL POKEMON
      case GET_POKEMON_DETAIL_START:
        draft.pokemonDetail.isLoading = true;
        break;
      case GET_POKEMON_DETAIL_SUCCESS:
        draft.pokemonDetail.isLoading = false;
        draft.pokemonDetail.data = action.data;
        break;
      case GET_POKEMON_DETAIL_FAILED:
        draft.pokemonDetail.isLoading = false;
        draft.pokemonDetail.data = [];
        draft.pokemonDetail.error = action.error;
        break;

      // GET LIST POKEMON TYPES
      case GET_POKEMON_TYPES_START:
        draft.pokemonTypes.isLoading = true;
        break;
      case GET_POKEMON_TYPES_SUCCESS:
        draft.pokemonTypes.isLoading = false;
        draft.pokemonTypes.data = action.data.results;
        break;
      case GET_POKEMON_TYPES_FAILED:
        draft.pokemonTypes.isLoading = false;
        draft.pokemonTypes.data = [];
        draft.pokemonTypes.error = action.error;
        break;
    }
  });

export default pokemonHomePageReducer;
