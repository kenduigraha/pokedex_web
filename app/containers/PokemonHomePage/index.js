/**
 *
 * PokemonHomePage
 *
 */

import React, { memo } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Helmet } from 'react-helmet';
import { FormattedMessage } from 'react-intl';
import { createStructuredSelector } from 'reselect';
import { compose } from 'redux';

import { useInjectSaga } from 'utils/injectSaga';
import { useInjectReducer } from 'utils/injectReducer';
import makeSelectPokemonHomePage from './selectors';
import reducer from './reducer';
import saga from './saga';
import messages from './messages';

export function PokemonHomePage() {
  useInjectReducer({ key: 'pokemonHomePage', reducer });
  useInjectSaga({ key: 'pokemonHomePage', saga });

  return (
    <div>
      <Helmet>
        <title>PokemonHomePage</title>
        <meta name="description" content="Description of PokemonHomePage" />
      </Helmet>
      <FormattedMessage {...messages.header} />
    </div>
  );
}

PokemonHomePage.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  pokemonHomePage: makeSelectPokemonHomePage(),
});

function mapDispatchToProps(dispatch) {
  return {
    dispatch,
  };
}

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export default compose(
  withConnect,
  memo,
)(PokemonHomePage);
