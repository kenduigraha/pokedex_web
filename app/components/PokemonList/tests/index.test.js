/**
 *
 * Tests for PokemonList
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
// import 'jest-dom/extend-expect'; // add some helpful assertions

import PokemonList from '../index';
import { DEFAULT_LOCALE } from '../../../i18n';
import { initialState } from '../../../containers/PokemonHomePage/reducer';

describe('<PokemonList />', () => {
  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <PokemonList
          pokemonList={{ ...initialState.pokemonList }}
          getPokemonList={() => {}}
          updateFlagInfinity={() => {}}
        />
      </IntlProvider>,
    );
    expect(spy).not.toHaveBeenCalled();
  });

  // it('Expect to have additional unit tests specified', () => {
  //   expect(true).toEqual(false);
  // });

  /**
   * Unskip this test to use it
   *
   * @see {@link https://jestjs.io/docs/en/api#testskipname-fn}
   */
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale={DEFAULT_LOCALE}>
        <PokemonList
          pokemonList={{ ...initialState.pokemonList }}
          getPokemonList={() => {}}
          updateFlagInfinity={() => {}}
        />
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
