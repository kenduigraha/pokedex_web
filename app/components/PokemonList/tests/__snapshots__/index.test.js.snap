// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonList /> Should render and match the snapshot 1`] = `
<div
  class="styles__PokemonListDiv-sc-1b5lzlw-0 cJsYwM"
  id="content"
>
  <div>
    <div
      class="ant-row"
      style="margin-left: -12px; margin-right: -12px;"
    />
  </div>
</div>
`;
